File: hardware/stepper_motor.py
"""
ESCO stepper motor driver over serial.
Port of APS2/Hardware/StepperMotor.m.

Units convention matches MATLAB: ConversionFactor is "units per step" where
units are mm for linear axes and degrees for rotation.
"""

from __future__ import annotations
import threading
import time
import serial

_LOG_PATH = "motor_debug.log"

def _log(msg: str):
    with open(_LOG_PATH, "a") as f:
        f.write(msg + "\n")


class StepperMotor:
    BAUD = 115200
    TIMEOUT = 2.0

    def __init__(self, name: str, port: str, address: int = 1):
        self.name = name
        self.port = port
        self.address = address

        # Calibration (set by StageController after construction).
        self.conversion_factor: float = 1.0   # units per step
        self.encoder_ratio: float = 1.0

        # Configurable hardware settings (applied via apply_settings()).
        self.motor_current: int = 3    # 3 = 1A
        self.hold_current: int = 0     # 0 = 0% (documented as 0% in new MATLAB default)
        self.microstepping: int = 0    # 0 = 256
        self.velocity: int = 100000
        self.acceleration: int = 5000
        self.encoder_type: int = 2     # 2 = Quadrature
        self.limit_lower: int = 0
        self.limit_upper: int = 0
        self.home_switch: int = 0
        self.homing_dir: int = 0       # 0=Pos, 1=Neg

        self.debug   = False          # set True to print raw serial traffic
        self._lock   = threading.Lock()
        self._serial = serial.Serial(port=port, baudrate=self.BAUD, timeout=self.TIMEOUT)
        time.sleep(0.2)
        print(f"--- Connecting to {name} ({port})...", end="", flush=True)

        idn = self.get_id()
        if not idn:
            self._serial.close()
            raise RuntimeError(f"Device {name} not responding.")

        if self.address >= 0:
            actual = self.get_address()
            if actual != self.address:
                print(f" WARNING: address mismatch (expected {self.address}, got {actual})")
            else:
                print(f" Connected (Addr: {actual}).")
        else:
            print(" Connected (address check skipped).")

        self.drive_on()

    def close(self):
        if self._serial.is_open:
            try:
                self.stop_movement()
                self.drive_off()
            except Exception:
                pass
            self._serial.close()
            print(f"--- Port {self.port} ({self.name}) closed.")

    # ------------------------------------------------------------------
    # Configuration
    # ------------------------------------------------------------------

    def apply_settings(self):
        try:
            self.set_setting("motor_current",  self.motor_current)
            self.set_setting("hold_current",   self.hold_current)
            self.set_setting("microstepping",  self.microstepping)
            self.set_setting("velocity",       self.velocity)
            self.set_setting("acceleration",   self.acceleration)
            self.set_setting("encoder_type",   self.encoder_type)
            self.set_setting("set_es_lower",   self.limit_lower)
            self.set_setting("set_es_upper",   self.limit_upper)
            self.set_setting("set_home",       self.home_switch)
            print(f"--- Settings applied to {self.name}.")
        except Exception as e:
            print(f"WARNING: failed to apply settings to {self.name}: {e}")

    # ------------------------------------------------------------------
    # Calibrated movement (mm for linear axes, deg for rotation)
    # ------------------------------------------------------------------

    def move_relative_units(self, value: float):
        # conv = mm/encoder_count; move_relative takes motor steps.
        # motor_steps = encoder_counts × |enc_ratio|
        encoder_counts = value / self.conversion_factor
        motor_steps = round(encoder_counts * abs(self.encoder_ratio))
        if self.debug:
            _log(f"[{self.name}] move_relative_units({value}) -> enc={encoder_counts:.1f} -> motor_steps={motor_steps}")
        self.move_relative(motor_steps)

    def move_absolute_units(self, value: float):
        encoder_counts = value / self.conversion_factor
        motor_steps = round(encoder_counts * abs(self.encoder_ratio))
        if self.debug:
            _log(f"[{self.name}] move_absolute_units({value}) -> enc={encoder_counts:.1f} -> motor_steps={motor_steps}")
        self.move_absolute(motor_steps)

    # Aliases for semantic clarity (matching MATLAB names).
    def move_relative_mms(self, mms: float):     self.move_relative_units(mms)
    def move_absolute_mms(self, mms: float):     self.move_absolute_units(mms)
    def move_relative_degrees(self, deg: float): self.move_relative_units(deg)
    def move_absolute_degrees(self, deg: float): self.move_absolute_units(deg)

    # ------------------------------------------------------------------
    # Status
    # ------------------------------------------------------------------

    def get_status_parsed(self) -> dict:
        resp = self._query("get_status")
        val = int(resp) if resp.lstrip("-").isdigit() else 0
        return {
            "raw":              val,
            "undervoltage":     bool(val & (1 << 0)),
            "comm_error":       bool(val & (1 << 1)),
            "moving":           bool(val & (1 << 2)),
            "moving_dir":       bool(val & (1 << 3)),
            "motor_error":      bool(val & (1 << 4)),
            "overtemp_pre":     bool(val & (1 << 5)),
            "overtemp":         bool(val & (1 << 6)),
            "open_cable":       bool(val & (1 << 7)),
            "short_circuit":    bool(val & (1 << 8)),
            "stall":            bool(val & (1 << 9)),
            "upper_limit_hit":  bool(val & (1 << 10)),
            "lower_limit_hit":  bool(val & (1 << 11)),
            "home_switch_hit":  bool(val & (1 << 12)),
        }

    def is_moving(self) -> bool:
        return self.get_status_parsed()["moving"]

    def wait_until_done(self, timeout: float = 10.0):
        # Small delay so the firmware has time to set the "moving" status bit
        # before we start polling — avoids a false-idle read right after a command.
        time.sleep(0.08)
        deadline = time.monotonic() + timeout
        while time.monotonic() < deadline:
            if not self.is_moving():
                return
            time.sleep(0.05)

    def get_position(self) -> float:
        """Return current position in calibrated units (mm or degrees)."""
        resp = self._query("get_position")
        steps = int(resp) if resp.lstrip("-").isdigit() else 0
        return steps * self.conversion_factor

    # ------------------------------------------------------------------
    # Basic commands
    # ------------------------------------------------------------------

    def get_id(self) -> str:
        return self._query("*IDN?")

    def get_address(self) -> int:
        resp = self._query("get_address")
        return int(resp) if resp.lstrip("-").isdigit() else -1

    def get_version(self) -> str:
        return self._query("get_version")

    def get_voltages(self) -> str:
        return self._query("get_voltages")

    def move_relative(self, steps: int):
        self._send(f"move_relative {int(steps)}")

    def move_absolute(self, steps: int):
        self._send(f"move_absolute {int(steps)}")

    def move_velocity(self, direction):
        if isinstance(direction, str):
            cmd = "const_v-" if direction.lower().startswith("n") or direction == "-" else "const_v+"
        else:
            cmd = "const_v-" if direction == 1 else "const_v+"
        self._send(cmd)

    def home(self, direction: int | None = None):
        if direction is None:
            direction = self.homing_dir
        self._send(f"home {int(direction)}")

    def stop_movement(self):
        self._send("stop_movement")

    def zero_position(self):
        self._send("zero_position")

    def set_output(self, state: int):
        self._query(f"set_output1 {int(state)}")

    def drive_on(self):
        self._send("drive_on")

    def drive_off(self):
        self._send("drive_off")

    def set_setting(self, name: str, value):
        self._send(f"set_setting {name} {int(value)}")

    def get_setting(self, name: str) -> float:
        resp = self._query(f"get_setting {name}")
        try:
            return float(resp)
        except ValueError:
            return float("nan")

    # ------------------------------------------------------------------
    # Serial helpers
    # ------------------------------------------------------------------

    def _query(self, cmd: str) -> str:
        with self._lock:
            self._serial.reset_input_buffer()
            self._serial.write((cmd + "\r\n").encode())
            raw = self._serial.readline()
            resp = raw.decode(errors="replace").strip()
            if self.debug:
                _log(f"[{self.name}] Q: {cmd!r}  RAW: {raw!r}  RESP: {resp!r}")
            if resp.startswith(cmd):
                resp = resp[len(cmd):].strip()
            return resp

    def _send(self, cmd: str):
        with self._lock:
            self._serial.reset_input_buffer()
            self._serial.write((cmd + "\r\n").encode())
            time.sleep(0.05)
            if self._serial.in_waiting:
                ack = self._serial.readline()
                if self.debug:
                    _log(f"[{self.name}] S: {cmd!r}  ACK: {ack!r}")

File: APS2/Hardware/StepperMotor.m
classdef StepperMotor < handle
    % StepperMotor - Improved Controller for ESCO Motors
    % Includes unit conversion, address verification, and status parsing.
    
    properties (Access = public)
        PortName string
        Address (1,1) double = 1 % Default Address
        Name string = "Motor"
        ConversionFactor double = 1.0; % Units per Step (mm or degrees)
        EncoderRatio double = 1.0; % Encoder Ratio
    end
    
    % Configurable Hardware Settings
    properties (Access = public)
        MotorCurrent    double = 3;  % 3 = 1A
        HoldCurrent     double = 0;  % 0 = 0%
        Microstepping   double = 0;  % 0 = 256
        Velocity        double = 100000;
        Acceleration    double = 5000;
        EncoderType     double = 2;  % 2 = Quadrature
        LimitLower      double = 0;  % 0 = None
        LimitUpper      double = 0;  % 0 = None
        HomeSwitch      double = 0;  % 0 = None
        HomingDir       double = 0;  % 0=Pos, 1=Neg
    end

    properties (Access = private)
        serialObj
        terminator = 'CR/LF';
        baudRate = 115200;
        timeout = 2;
    end

    methods
        function obj = StepperMotor(varargin)
            % Constructor supports two modes:
            % 1. obj = StepperMotor(comPort)  -> Legacy/Simple mode
            % 2. obj = StepperMotor(name, comPort, address) -> Full mode
            
            if nargin == 1
                % Legacy mode: StepperMotor('COM3')
                obj.Name = "Motor";
                obj.PortName = varargin{1};
                obj.Address = 1; % Default
            elseif nargin == 3
                % Full mode: StepperMotor("X", "COM7", 1)
                obj.Name = varargin{1};
                obj.PortName = varargin{2};
                obj.Address = varargin{3};
            else
                error('Invalid arguments. Use StepperMotor(port) or StepperMotor(name, port, address).');
            end

            try
                obj.serialObj = serialport(obj.PortName, obj.baudRate);
                obj.serialObj.Timeout = obj.timeout;
                configureTerminator(obj.serialObj, obj.terminator);
                flush(obj.serialObj);
            catch ME
                error('Failed to open %s for %s. Check connection.', obj.PortName, obj.Name);
            end
            
            pause(0.2);
            fprintf('--- Connecting to %s (%s)...', obj.Name, obj.PortName);
            
            % Check Communication & Address
            idn = obj.get_id();
            if isempty(idn)
                delete(obj.serialObj);
                error(' Device %s not responding.', obj.Name);
            end
            
            % Verify Address
            if obj.Address >= 0
                addr = obj.get_address();
                if addr ~= obj.Address
                    warning('Address Mismatch on %s! Expected %d, got %d.', obj.Name, obj.Address, addr);
                else
                    fprintf(' Connected (Addr: %d).\n', addr);
                end
            else
                fprintf(' Connected (Address check skipped).\n');
            end
            
            % Auto-Enable Drive (As requested)
            obj.drive_on();
        end

        function delete(obj)
            if ~isempty(obj.serialObj) && isvalid(obj.serialObj)
                try
                    obj.stop_movement(); 
                    obj.drive_off(); 
                catch
                end
                delete(obj.serialObj);
                fprintf('--- Port %s (%s) closed.\n', obj.PortName, obj.Name);
            end
        end

        % --- Configuration ---
        function apply_settings(obj)
            try
                obj.set_setting('motor_current', obj.MotorCurrent);
                obj.set_setting('hold_current',  obj.HoldCurrent);
                obj.set_setting('microstepping', obj.Microstepping);
                obj.set_setting('velocity',      obj.Velocity);
                obj.set_setting('acceleration',  obj.Acceleration);
                obj.set_setting('encoder_type',  obj.EncoderType);
                obj.set_setting('set_es_lower',  obj.LimitLower);
                obj.set_setting('set_es_upper',  obj.LimitUpper);
                obj.set_setting('set_home',      obj.HomeSwitch);
                
                % Note: EncoderRatio setting requires specific firmware support command
                % obj.set_setting('encoder_ratio', obj.EncoderRatio);
                
                fprintf('--- Settings applied to %s.\n', obj.Name);
            catch ME
                warning('Failed to apply settings to %s: %s', obj.Name, ME.message);
            end
        end

        % --- Movement (Calibrated Units) ---
        
        % Primary Methods (Millimeters)
        function move_relative_mms(obj, mms)
            % Converts mm to steps using ConversionFactor
            steps = round(mms / obj.ConversionFactor);
            obj.move_relative(steps);
        end

        function move_absolute_mms(obj, mms)
            % Converts mm to steps using ConversionFactor
            steps = round(mms / obj.ConversionFactor);
            obj.move_absolute(steps);
        end
        
        % Aliases for Rotation (Degrees)
        % These do the exact same math, but allow semantic clarity for Rot stage
        function move_relative_degrees(obj, deg)
            steps = round(deg / obj.ConversionFactor);
            obj.move_relative(steps);
        end
        
        function move_absolute_degrees(obj, deg)
            steps = round(deg / obj.ConversionFactor);
            obj.move_absolute(steps);
        end
        
        % --- Basic Commands ---
        function idn = get_id(obj)
            idn = obj.sendCommand('*IDN?', 'query');
        end
        
        function addr = get_address(obj)
            resp = obj.sendCommand('get_address', 'query');
            addr = str2double(resp);
        end

        function ver = get_version(obj)
            ver = obj.sendCommand('get_version', 'query');
        end
        
        function volts = get_voltages(obj)
             volts = obj.sendCommand('get_voltages', 'query');
        end

        % --- Status ---
        function status = get_status_parsed(obj)
            resp = obj.sendCommand('get_status', 'query');
            val = str2double(resp);
            
            status = struct();
            status.Raw = val;
            status.Undervoltage       = bitget(val, 1);
            status.CommError          = bitget(val, 2);
            status.Moving             = bitget(val, 3);
            status.MovingDir          = bitget(val, 4);
            status.MotorError         = bitget(val, 5);
            status.OvertempPre        = bitget(val, 6);
            status.Overtemp           = bitget(val, 7);
            status.OpenCable          = bitget(val, 8);
            status.ShortCircuit       = bitget(val, 9);
            status.Stall              = bitget(val, 10);
            status.UpperLimitHit      = bitget(val, 11);
            status.LowerLimitHit      = bitget(val, 12);
            status.HomeSwitchHit      = bitget(val, 13);
        end

        function pos = get_position(obj)
            resp = obj.sendCommand('get_position', 'query');
            steps = str2double(resp);
            % Return position in calibrated UNITS (mm or degrees)
            pos = steps * obj.ConversionFactor;
        end

        % --- Hardware Commands ---
        function move_absolute(obj, steps)
            obj.sendCommand(sprintf('move_absolute %d', steps), 'set');
        end

        function move_relative(obj, steps)
            obj.sendCommand(sprintf('move_relative %d', steps), 'set');
        end
        
        function move_velocity(obj, direction)
            if (ischar(direction) || isstring(direction))
                if strcmpi(direction, 'negative') || direction == "-"
                    cmd = "const_v-";
                else
                    cmd = "const_v+";
                end
            else
                if direction == 1
                    cmd = "const_v-";
                else
                    cmd = "const_v+";
                end
            end
            obj.sendCommand(cmd, 'set');
        end

        function home(obj, direction)
            if nargin < 2
                direction = obj.HomingDir;
            end
            obj.sendCommand(sprintf('home %d', direction), 'set');
        end

        function stop_movement(obj)
            obj.sendCommand('stop_movement', 'set');
        end
        
        function zero_position(obj)
            obj.sendCommand('zero_position', 'set');
        end
        
        function set_output(obj, state)
            obj.sendCommand(sprintf('set_output1 %d', state), 'query'); 
        end

        % --- System ---
        function drive_on(obj)
            obj.sendCommand('drive_on', 'set');
        end

        function drive_off(obj)
            obj.sendCommand('drive_off', 'set');
        end
        
        function set_setting(obj, name, val)
            obj.sendCommand(sprintf('set_setting %s %d', name, val), 'set');
        end

        function val = get_setting(obj, name)
            resp = obj.sendCommand(sprintf('get_setting %s', name), 'query');
            val = str2double(resp);
        end
    end
    
    methods (Access = private)
        function ok = check_communication(obj)
            resp = obj.sendCommand('*IDN?', 'query', true);
            ok = ~isempty(resp) && ~contains(resp, 'Error');
        end

        function response = sendCommand(obj, cmd, type, silent)
            if nargin < 4, silent = false; end
            
            response = '';
            try
                flush(obj.serialObj);
                writeline(obj.serialObj, cmd);
                
                if strcmpi(type, 'query')
                    response = readline(obj.serialObj);
                    response = strtrim(response);
                    if startsWith(response, cmd), response = extractAfter(response, cmd); end
                    response = strtrim(response);
                else
                    pause(0.05); 
                    if obj.serialObj.NumBytesAvailable > 0
                        readline(obj.serialObj);
                    end
                end
            catch ME
                if ~silent
                    warning('Serial Error on "%s": %s', cmd, ME.message);
                end
            end
        end
    end
end